pkg/registry/batch/job/strategy_test.go

```diff
@@ -709,44 +815,205 @@ func TestJobStrategyWithGeneration(t *testing.T) {
 		},
 		Spec: batch.JobSpec{
 			Selector: nil,
-			Template: validPodTemplateSpec,
-		},
+			Template: api.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: validSelector.MatchLabels,
+				},
+				Spec: api.PodSpec{
+					RestartPolicy: api.RestartPolicyOnFailure,
+					DNSPolicy:     api.DNSClusterFirst,
+					Containers:    []api.Container{{Name: "abc", Image: "image", ImagePullPolicy: "IfNotPresent", TerminationMessagePolicy: api.TerminationMessageReadFile}},
+				},
+			}},
 	}
 
-	Strategy.PrepareForCreate(ctx, job)
-	errs := Strategy.Validate(ctx, job)
-	if len(errs) != 0 {
-		t.Errorf("Unexpected error validating %v", errs)
+	testcases := map[string]struct {
+		jobName  string
+		warnings string
+	}{
+		"happy path job": {
+			jobName: "myjob2",
+		},
+		"dns invalid name": {
+			warnings: "a DNS label is recommended",
+			jobName:  "my job2",
+		},
+	}
+	for name, tc := range testcases {
+		t.Run(name, func(t *testing.T) {
+			job.Name = tc.jobName
+			actualWarnings := Strategy.WarningsOnCreate(ctx, job)
+			if tc.warnings != "" {
+				for _, val := range actualWarnings {
+					if !strings.Contains(val, tc.warnings) {
+						t.Errorf("Detected error %s that is different from %s", val, tc.warnings)
+					}
+				}
+			} else {
+				if len(actualWarnings) != 0 {
+					t.Errorf("Do not expect warnings on create but got them")
+				}
+			}
+		})
 	}
+}
+func TestJobStrategy_Validate(t *testing.T) {
+	ctx := genericapirequest.NewDefaultContext()
 
-	// Validate the stuff that validation should have validated.
-	if job.Spec.Selector == nil {
-		t.Errorf("Selector not generated")
+	theUID := types.UID("1a2b3c4d5e6f7g8h9i0k")
+	validSelector := &metav1.LabelSelector{
+		MatchLabels: map[string]string{"a": "b"},
 	}
-	expectedLabels := make(map[string]string)
-	expectedLabels["controller-uid"] = string(theUID)
-	if !reflect.DeepEqual(job.Spec.Selector.MatchLabels, expectedLabels) {
-		t.Errorf("Expected label selector not generated")
+
+	testcases := map[string]struct {
+		job                *batch.Job
+		dontgeneratelabels bool
```

Comment:
I suggest to rename it to avoid double negations in the asserts, such as `wantGeneratedLabels`. However, maybe we could just have `wantJob` and compare the entire objects using `cmp.Diff`. It would allow more generic testing as the test method is also generic `TestJobStrategy_Validate`.

Reply:
Yea I think that is probably better.  I'll look into that.

Reply:
TBH I was aiming to keep the logic the same as we did in the original code.  But I do think this suggestion is better.